### PR TITLE
ADD vaapi encoder

### PR DESCRIPTION
**Pull request by xhejtman**
added vaapi encoder properly

### CODE REVIEW

**Review by ehfd — APPROVED** [review]
Looks good to me.

**Comment by ehfd** [comment]
Resolves part of #34.

File: src/selkies_gstreamer/gstwebrtc_app.py [hunk 1]
```diff
@@ -188,6 +188,19 @@ def build_video_pipeline(self):
...
+            convert_caps = Gst.caps_from_string("video/x-raw")
+            convert_caps.set_value("format", "NV12")
+            convert_capsfilter = Gst.ElementFactory.make("capsfilter")
+            convert_capsfilter.set_property("caps", convert_caps)
+            self.pipeline.add(convert)
+            self.pipeline.add(convert_capsfilter)
+            if not Gst.Element.link(ximagesrc_capsfilter, convert):
+                raise GSTWebRTCAppError("Failed to link ximagesrc_capsfilter -> convert")
+            if not Gst.Element.link(convert, convert_capsfilter):
+                raise GSTWebRTCAppError("Failed to link convert -> convert_capsfilter")
```
Comment (danisla) [inline src/selkies_gstreamer/gstwebrtc_app.py:202]: replace `convert` with `vaapipostproc`

**Review by danisla — CHANGES_REQUESTED** [review]

File: src/selkies_gstreamer/gstwebrtc_app.py (hunk 1 above)
Comment (ehfd) [reply src/selkies_gstreamer/gstwebrtc_app.py]: Confirmed that it has been edited.

File: src/selkies_gstreamer/gstwebrtc_app.py [hunk 2]
```diff
@@ -188,6 +188,19 @@ def build_video_pipeline(self):
...
+        if self.encoder.startswith("vaapi"):
+            convert = Gst.ElementFactory.make("vaapipostproc")
+            convert_caps = Gst.caps_from_string("video/x-raw")
+            convert_caps.set_value("format", "NV12")
+            convert_capsfilter = Gst.ElementFactory.make("capsfilter")
+            convert_capsfilter.set_property("caps", convert_caps)
+            self.pipeline.add(convert)
+            self.pipeline.add(convert_capsfilter)
+            if not Gst.Element.link(ximagesrc_capsfilter, convert):
+                raise GSTWebRTCAppError("Failed to link ximagesrc_capsfilter -> convert")
```
[the remark this answers is not in the document]
Comment (ehfd) [reply src/selkies_gstreamer/gstwebrtc_app.py]: Confirmed that it has been edited.

**Comment by ehfd** [comment]
gst-vaapi is required to be included in the build procedures. @danisla Please keep it in mind.

**Comment by danisla** [comment]
@ehfd are there any changes to the gstreamer source compilation needed to get vaapi support? What are the runtime libs required?

**Review by danisla — COMMENTED** [review]
What are the requirements to get the `vaapih264enc` elements? Are they currently compiled with the gstreamer base image, or are there additional compile-time or runtime dependencies?

**Comment by ehfd** [comment]
@danisla GST-VAAPI
Also note that only i965 is supported for Intel, not iHD

**Comment by danisla** [comment]
> @danisla bundle GST-VAAPI (VA drivers required in host of course) Also note that only i965 is supported for Intel, not iHD, and this is experimental

I don't have an i965 to test with, so just to confirm, no changes to the `addons/gstreamer/Dockerfile` are required and this is purely a runtime dependency?

**Comment by ehfd** [comment]
@xhejtman

**Comment by ehfd** [comment]
@danisla I do not see it as an available plugin on GStreamer.

**Comment by danisla** [comment]
The gstreamer ubuntu20.04 [build logs](https://pipelines.actions.githubusercontent.com/serviceHosts/acc96cc8-eb83-4833-8f1f-aa0f322590c8/_apis/pipelines/1/runs/239/signedlogcontent/9?urlExpires=2022-09-09T04%3A47%3A06.3147955Z&urlSigningMethod=HMACV1&urlSignature=jKDeTpaeqXv80YaTOuM5Ji%2B4QqUN4swqeutE3G30ctQ%3D) shows that the `gstreamer-vaapi` subproject is not being built. 

```
2022-08-19T16:10:40.1095173Z Subproject gstreamer-vaapi : skipped: feature vaapi disabled
```

Can someone please propose and test a change to the `addons/gstreamer/Dockerfile` to include vaapi support so that this PR will work as expected?

**Comment by ehfd** [comment]
Note: set `USE_DRM` for `gstreamer-vaapi/config.h` to `true`. This should be somehere in the build settings.
Dependencies:
```
libva_req = ['>= 0.39.0', '!= 0.99.0']
libwayland_req = '>= 1.11.0'
libdrm_req = '>= 2.4.98'
gst_req = '>= @0@.@1@.0'.format(gst_version_major, gst_version_minor)
```

**Comment by ehfd** [comment]
@danisla libva, libwayland, libdrm

**Comment by xhejtman** [comment]
hmm, yes, to build vaapi subproject was painful. I built it manualy. Also note, for runtime, i965-va-driver-shaders is needed.

**Comment by ehfd** [comment]
Should we stop here for now, perhaps?
`vah264enc` is in master and should be in GST 1.21. It should have a better experience.

**Comment by xhejtman** [comment]
It makes sense, it was really not easy to build it properly.

**Comment by ehfd** [comment]
I think this PR is ready to be revived, now that `vah264enc` is in 1.21.1.

**Comment by ehfd** [comment]
@xhejtman The vaapi issue with the framerate last time might have been with `MIT-SHM` missing, btw.

**Comment by danisla** [comment]
I've been testing with gstreamer 1.21.1, and `vah264enc` is present and is working with gstreamer standalone like this:
```
gst-launch-1.0 ximagesrc ! vapostproc ! video/x-raw,format=NV12 ! vah264enc ! fakesink
```

However, regardless of encoder setting, gstreamer segfaults when running with webrtcbin on gst-1.21.1. Also I am on the fence for supporting non-stable gstreamer releases (odd numbered releases) because of issues like this..,

**Comment by xhejtman** [comment]
which intel driver are you using? It works with shader only. Also, vapostproc is problematic. I would start either with SW/CPU video convert. And perhaps use different intel driver. iHD is mostly buggy.

**Comment by danisla** [comment]
@xhejtman the vaapi codecs and vapostproc seemed to work just fine, I don't think the issue was with the vaapi plugin, but rather with webrtc in gstreamer v1.21.1. Like I said, the x264enc pipeline doesn't work on 1.21.1 either. 

I'm probably just going to stick with the stable releases of gstreamer since odd releases are not stable.

**Comment by ehfd** [comment]
Relevant? @danisla :
https://gitlab.freedesktop.org/gstreamer/gstreamer/-/commit/430ec0d860738199b81c4aac73a6b057918be144

I think we also need to check if dependencies have changed in https://github.com/selkies-project/selkies-gstreamer/commit/276fe41d4a23d0985b819e3fc64b173753293456.

`-Dvaapi=enabled -Dvaapi:encoders=true -Dvaapi:drm=true`
Moreover, is this not for the VAAPI plugin, not the VA plugin within bad?

**Comment by danisla** [comment]
@xhejtman I'm having trouble getting this to work with 1.20.4. 

This is what my vainfo output looks like:

```
libva info: VA-API version 1.7.0
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so
libva info: Found init function __vaDriverInit_1_7
libva error: /usr/lib/x86_64-linux-gnu/dri/iHD_drv_video.so init failed
libva info: va_openDriver() returns 1
libva info: Trying to open /usr/lib/x86_64-linux-gnu/dri/i965_drv_video.so
libva info: Found init function __vaDriverInit_1_6
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.7 (libva 2.6.0)
vainfo: Driver version: Intel i965 driver for Intel(R) Bay Trail - 2.4.0
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Simple            : VAEntrypointEncSlice
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointEncSlice
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointEncSlice
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointEncSlice
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointEncSlice
      VAProfileH264StereoHigh         : VAEntrypointVLD
      VAProfileVC1Simple              : VAEntrypointVLD
      VAProfileVC1Main                : VAEntrypointVLD
      VAProfileVC1Advanced            : VAEntrypointVLD
      VAProfileNone                   : VAEntrypointVideoProc
      VAProfileJPEGBaseline           : VAEntrypointVLD
```

then `gst-inspect-1.0 vaapi` shows this:

```
0:00:00.656034240    25 0x564effb1a200 WARN                 default gstvaapi.c:231:plugin_init: Cannot create a VA display
Plugin Details:
  Name                     vaapi
  Description              VA-API based elements
  Filename                 /opt/gstreamer/lib/x86_64-linux-gnu/gstreamer-1.0/libgstvaapi.so
  Version                  1.20.4
  License                  LGPL
  Source module            gstreamer-vaapi
  Source release date      2022-10-12
  Binary package           gstreamer-vaapi
  Origin URL               Unknown package origin


  0 features:
```

**Comment by ehfd** [comment]
@danisla First thing. Is /dev/dri/cardX from an Intel GPU accessible?

**Comment by ehfd** [comment]
https://blogs.igalia.com/vjaquez/2018/03/28/gstreamer-va-api-troubleshooting/

**Comment by danisla** [comment]
> @danisla First thing. Is /dev/dri/cardX from an Intel GPU accessible?

yes, as you can see from my previous comment, the `vainfo` output is good, permissions are good on /dev/dri.

**Comment by ehfd** [comment]
I'll try to fiddle with the build process now. Though I don't have a Linux node with Intel right now.

**Comment by ehfd** [comment]
[gst1-21-1.txt](https://github.com/selkies-project/selkies-gstreamer/files/9805471/gst1-21-1.txt)
What I see so far. Without `-Dvaapi=enabled -Dvaapi:encoders=true -Dvaapi:drm=true` so all these have nothing to do with `va` in gst-bad.

**Comment by danisla** [comment]
I would like to see this working with 1.20.4 if possible. I am assuming that is what @xhejtman got it working with.

We control the gstreamer build so we should be able to make it work if we know the build steps.

**Comment by ehfd** [comment]
I also see a segfault on 1.21.1.

I believe it might be related to https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1508, although not limited to NVIDIA

**Comment by danisla** [comment]
> I also see a segfault on 1.21.1.
> 
> Edit: It happens right here
> 
> ```
> 0:00:34.057128636 45609      0x24cc980 DEBUG               GST_PADS gstpad.c:4193:gst_pad_query:<capsfilter2:sink> sent query 0x1942cf0 (caps), result 1
> 0:00:34.057139942 45609      0x24cc980 DEBUG               GST_CAPS gstpad.c:2837:gst_pad_get_allowed_caps:<nvenc:src> allowed caps video/x-h264, width=(int)[ 33, 4096 ], height=(int)[ 17, 4096 ], framerate=(fraction)[ 0/1, 2147483647/1 ], stream-format=(string)byte-stream, alignment=(string)au, profile=(string)high
> 0:00:34.064355940 45609      0x24cc980 INFO                   nvenc gstnvbaseenc.c:1864:gst_nv_base_enc_set_format:<nvenc> configured encoder
> 0:00:34.064378957 45609      0x24cc980 DEBUG                  nvenc gstnvbaseenc.c:1624:gst_nv_base_enc_calculate_num_prealloc_buffers:<nvenc> Calculated num buffers: 5 (lookahead 0, frameIntervalP 1)
> 0:00:34.067076449 45609      0x24cc980 INFO                   nvenc gstnvbaseenc.c:1959:gst_nv_base_enc_set_format:<nvenc> allocated output buffer  0: 0x7f708008fe10
> 0:00:34.068826456 45609      0x24cc980 INFO                   nvenc gstnvbaseenc.c:1959:gst_nv_base_enc_set_format:<nvenc> allocated output buffer  1: 0x7f7080005550
> 0:00:34.070164906 45609      0x24cc980 INFO                   nvenc gstnvbaseenc.c:1959:gst_nv_base_enc_set_format:<nvenc> allocated output buffer  2: 0x7f70804e8180
> 0:00:34.071452229 45609      0x24cc980 INFO                   nvenc gstnvbaseenc.c:1959:gst_nv_base_enc_set_format:<nvenc> allocated output buffer  3: 0x7f708017a610
> /etc/selkies-gstreamer-entrypoint.sh: line 57: 45609 Segmentation fault      (core dumped) selkies-gstreamer --json_config="${SELKIES_USER_CONFIG_FILE}" --addr="0.0.0.0" --port="8080" $@
> ```
> 
> I believe it might be https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1508.

I don't think it's related to the encoder, it segfaults for me with any encoder.

See my previous comment:

> However, regardless of encoder setting, gstreamer segfaults when running with webrtcbin on gst-1.21.1. Also I am on the fence for supporting non-stable gstreamer releases (odd numbered releases) because of issues like this..,

**Comment by ehfd** [comment]
> However, regardless of encoder setting, gstreamer segfaults when running with webrtcbin on gst-1.21.1. Also I am on the fence for supporting non-stable gstreamer releases (odd numbered releases) because of issues like this..,

I understand the odd numbered releases are unreliable, but I do still need to determine if it's on our side or it's the code. On second look, it happens right when the SDP connection is established on any encoder, yes.

**Comment by ehfd** [comment]
@danisla 
Strangely, x264 and vp8 works with gst 1.21.1 after I did something outside of my understanding.
Alhough logs are spamming with `0:01:04.799529049 77363 0x7f716c0a24c0 WARN              rtpsession gstrtpsession.c:2438:gst_rtp_session_chain_send_rtp_common:<rtpsession0> Can't determine running time for this packet without knowing configured latency`

Let me refresh my container and setup again.

Edit: I definitely did not confuse versions. They work in a certain condition, but is inconsistent... Need to find what happened.

**Comment by ehfd** [comment]
Overall, however, the behavior of 1.21 is too puzzling to be used for Selkies.

**Comment by xhejtman** [comment]
Do you have i965-va-driver or i965-va-driver-shaders?

**Comment by ehfd** [comment]
@xhejtman Check https://github.com/selkies-project/selkies-gstreamer/blob/gst-1.21.1/Dockerfile.example and https://github.com/selkies-project/selkies-gstreamer/blob/gst-1.21.1/addons/gstreamer/Dockerfile.

Overall: https://github.com/selkies-project/selkies-gstreamer/commit/276fe41d4a23d0985b819e3fc64b173753293456
Do you mean we need them while building GST? I didn't think so.

**Comment by xhejtman** [comment]
so, it looks like you have the shaders version. strange, so it might be related to particular intel gpu, some might work, some might not.

**Comment by danisla** [comment]
I got it working with gstreamer 1.20, was missing a build time dependency of `libudev-dev`

I'll create a new PR to update to the gstreamer 1.20 branch with vaapi support.

**Comment by ehfd** [comment]
> I got it working with gstreamer 1.20, was missing a build time dependency of `libudev-dev`
> 
> I'll create a new PR to update to the gstreamer 1.20 branch with vaapi support.

Does it happen to work with the VA plugin (not VAAPI)? It probably needs a plugin rank change if in 1.20.

**Comment by danisla** [comment]
Only vaapi, the va plugin does not have the encoder elements in gstreamer 1.20, we'll have to wait until 1.22 to support the new va plugin.

**Comment by ehfd** [comment]
Please change the docs too if there are runtime dependency changes.
There should be 2 places that should be changed.

**Comment by ehfd** [comment]
@danisla @xhejtman 

Issue with Selkies on GST 1.21.x raised on https://gitlab.freedesktop.org/gstreamer/gstreamer/-/issues/1605

**Comment by ehfd** [comment]
Implemented in https://github.com/selkies-project/selkies-gstreamer/commit/736292ca1ef7e1a705330d1b70f9d23bb09d9d0f

**Pull request closed without merging**